File: python/basic_operations.py
#!/usr/bin/env python3
"""
Exemplos básicos de operações com MinIO usando Python.
Este script demonstra operações fundamentais como criar buckets,
fazer upload, download e listar objetos.
"""

from minio import Minio
from minio.error import S3Error
import io
import os

# Configuração do cliente MinIO
client = Minio(
    "localhost:9000",
    access_key="minioadmin",
    secret_key="minioadmin",
    secure=False  # Use True se estiver usando HTTPS
)

def create_bucket(bucket_name):
    """
    Cria um novo bucket no MinIO.
    
    Args:
        bucket_name (str): Nome do bucket a ser criado
    """
    try:
        # Verifica se o bucket já existe
        if not client.bucket_exists(bucket_name):
            client.make_bucket(bucket_name)
            print(f"✓ Bucket '{bucket_name}' criado com sucesso!")
        else:
            print(f"✓ Bucket '{bucket_name}' já existe.")
    except S3Error as err:
        print(f"✗ Erro ao criar bucket: {err}")

def upload_file(bucket_name, object_name, file_path):
    """
    Faz upload de um arquivo para o MinIO.
    
    Args:
        bucket_name (str): Nome do bucket
        object_name (str): Nome do objeto no MinIO
        file_path (str): Caminho do arquivo local
    """
    try:
        client.fput_object(bucket_name, object_name, file_path)
        print(f"✓ Arquivo '{file_path}' enviado como '{object_name}' para o bucket '{bucket_name}'")
    except S3Error as err:
        print(f"✗ Erro no upload: {err}")

def upload_string(bucket_name, object_name, content):
    """
    Faz upload de uma string como objeto no MinIO.
    
    Args:
        bucket_name (str): Nome do bucket
        object_name (str): Nome do objeto no MinIO
        content (str): Conteúdo a ser enviado
    """
    try:
        # Converte string para bytes
        content_bytes = content.encode('utf-8')
        
        # Cria um objeto BytesIO
        content_stream = io.BytesIO(content_bytes)
        
        # Upload do stream
        client.put_object(
            bucket_name,
            object_name,
            content_stream,
            length=len(content_bytes),
            content_type='text/plain'
        )
        print(f"✓ Conteúdo enviado como '{object_name}' para o bucket '{bucket_name}'")
    except S3Error as err:
        print(f"✗ Erro no upload: {err}")

def download_file(bucket_name, object_name, file_path):
    """
    Faz download de um arquivo do MinIO.
    
    Args:
        bucket_name (str): Nome do bucket
        object_name (str): Nome do objeto no MinIO
        file_path (str): Caminho onde salvar o arquivo
    """
    try:
        client.fget_object(bucket_name, object_name, file_path)
        print(f"✓ Objeto '{object_name}' baixado para '{file_path}'")
    except S3Error as err:
        print(f"✗ Erro no download: {err}")

def list_objects(bucket_name):
    """
    Lista todos os objetos em um bucket.
    
    Args:
        bucket_name (str): Nome do bucket
    """
    try:
        print(f"\n📋 Objetos no bucket '{bucket_name}':")
        objects = client.list_objects(bucket_name)
        for obj in objects:
            print(f"  - {obj.object_name} (Tamanho: {obj.size} bytes)")
    except S3Error as err:
        print(f"✗ Erro ao listar objetos: {err}")

def delete_object(bucket_name, object_name):
    """
    Remove um objeto do bucket.
    
    Args:
        bucket_name (str): Nome do bucket
        object_name (str): Nome do objeto a ser removido
    """
    try:
        client.remove_object(bucket_name, object_name)
        print(f"✓ Objeto '{object_name}' removido do bucket '{bucket_name}'")
    except S3Error as err:
        print(f"✗ Erro ao remover objeto: {err}")

def main():
    """
    Função principal que executa todos os exemplos.
    """
    print("=" * 60)
    print("MinIO Python - Operações Básicas")
    print("=" * 60)
    
    bucket_name = "python-example-bucket"
    
    # 1. Criar bucket
    print("\n1. Criando bucket...")
    create_bucket(bucket_name)
    
    # 2. Upload de string
    print("\n2. Fazendo upload de conteúdo...")
    upload_string(bucket_name, "hello.txt", "Olá MinIO! Este é um teste do Python.")
    
    # 3. Upload de arquivo (criamos um arquivo temporário para o exemplo)
    print("\n3. Criando e fazendo upload de arquivo...")
    temp_file = "/tmp/example.txt"
    with open(temp_file, 'w') as f:
        f.write("Este é um arquivo de exemplo criado em Python.\n")
        f.write("MinIO é compatível com Amazon S3.\n")
    upload_file(bucket_name, "example.txt", temp_file)
    
    # 4. Listar objetos
    print("\n4. Listando objetos no bucket...")
    list_objects(bucket_name)
    
    # 5. Download de arquivo
    print("\n5. Fazendo download de arquivo...")
    download_path = "/tmp/downloaded_example.txt"
    download_file(bucket_name, "example.txt", download_path)
    
    # Exibir conteúdo do arquivo baixado
    with open(download_path, 'r') as f:
        content = f.read()
        print(f"   Conteúdo baixado:\n   {content}")
    
    # 6. Deletar objeto (opcional - descomente se quiser testar)
    # print("\n6. Removendo objeto...")
    # delete_object(bucket_name, "hello.txt")
    
    print("\n" + "=" * 60)
    print("✓ Todos os exemplos foram executados!")
    print("=" * 60)

if __name__ == "__main__":
    main()

File: nodejs/basicOperations.js
/**
 * Operações básicas com MinIO usando Node.js
 * Este script demonstra as operações fundamentais de integração com MinIO.
 */

const Minio = require('minio');
const fs = require('fs');
const path = require('path');

// Configuração do cliente MinIO
const minioClient = new Minio.Client({
    endPoint: 'localhost',
    port: 9000,
    useSSL: false,
    accessKey: 'minioadmin',
    secretKey: 'minioadmin'
});

/**
 * Cria um novo bucket
 * @param {string} bucketName - Nome do bucket
 */
async function createBucket(bucketName) {
    try {
        const exists = await minioClient.bucketExists(bucketName);
        if (!exists) {
            await minioClient.makeBucket(bucketName, 'us-east-1');
            console.log(`✓ Bucket '${bucketName}' criado com sucesso!`);
        } else {
            console.log(`✓ Bucket '${bucketName}' já existe.`);
        }
    } catch (err) {
        console.error(`✗ Erro ao criar bucket: ${err.message}`);
    }
}

/**
 * Faz upload de um arquivo para o MinIO
 * @param {string} bucketName - Nome do bucket
 * @param {string} objectName - Nome do objeto no MinIO
 * @param {string} filePath - Caminho do arquivo local
 */
async function uploadFile(bucketName, objectName, filePath) {
    try {
        const metaData = {
            'Content-Type': 'text/plain'
        };
        
        await minioClient.fPutObject(bucketName, objectName, filePath, metaData);
        console.log(`✓ Arquivo '${filePath}' enviado como '${objectName}'`);
    } catch (err) {
        console.error(`✗ Erro no upload: ${err.message}`);
    }
}

/**
 * Faz upload de conteúdo de string
 * @param {string} bucketName - Nome do bucket
 * @param {string} objectName - Nome do objeto
 * @param {string} content - Conteúdo a ser enviado
 */
async function uploadString(bucketName, objectName, content) {
    try {
        const buffer = Buffer.from(content, 'utf-8');
        const metaData = {
            'Content-Type': 'text/plain',
            'Content-Length': buffer.length
        };
        
        await minioClient.putObject(bucketName, objectName, buffer, metaData);
        console.log(`✓ Conteúdo enviado como '${objectName}'`);
    } catch (err) {
        console.error(`✗ Erro no upload: ${err.message}`);
    }
}

/**
 * Faz download de um arquivo do MinIO
 * @param {string} bucketName - Nome do bucket
 * @param {string} objectName - Nome do objeto
 * @param {string} filePath - Caminho onde salvar o arquivo
 */
async function downloadFile(bucketName, objectName, filePath) {
    try {
        await minioClient.fGetObject(bucketName, objectName, filePath);
        console.log(`✓ Objeto '${objectName}' baixado para '${filePath}'`);
    } catch (err) {
        console.error(`✗ Erro no download: ${err.message}`);
    }
}

/**
 * Lista todos os objetos em um bucket
 * @param {string} bucketName - Nome do bucket
 */
async function listObjects(bucketName) {
    try {
        console.log(`\n📋 Objetos no bucket '${bucketName}':`);
        const stream = minioClient.listObjects(bucketName, '', true);
        
        return new Promise((resolve, reject) => {
            const objects = [];
            stream.on('data', (obj) => {
                console.log(`  - ${obj.name} (Tamanho: ${obj.size} bytes)`);
                objects.push(obj);
            });
            stream.on('error', reject);
            stream.on('end', () => resolve(objects));
        });
    } catch (err) {
        console.error(`✗ Erro ao listar objetos: ${err.message}`);
    }
}

/**
 * Remove um objeto do bucket
 * @param {string} bucketName - Nome do bucket
 * @param {string} objectName - Nome do objeto
 */
async function deleteObject(bucketName, objectName) {
    try {
        await minioClient.removeObject(bucketName, objectName);
        console.log(`✓ Objeto '${objectName}' removido`);
    } catch (err) {
        console.error(`✗ Erro ao remover objeto: ${err.message}`);
    }
}

/**
 * Obtém informações sobre um objeto
 * @param {string} bucketName - Nome do bucket
 * @param {string} objectName - Nome do objeto
 */
async function getObjectInfo(bucketName, objectName) {
    try {
        const stat = await minioClient.statObject(bucketName, objectName);
        console.log(`\n📊 Informações do objeto '${objectName}':`);
        console.log(`  - Tamanho: ${stat.size} bytes`);
        console.log(`  - Tipo: ${stat.metaData['content-type']}`);
        console.log(`  - Última modificação: ${stat.lastModified}`);
        console.log(`  - ETag: ${stat.etag}`);
    } catch (err) {
        console.error(`✗ Erro ao obter informações: ${err.message}`);
    }
}

/**
 * Função principal
 */
async function main() {
    console.log('='.repeat(60));
    console.log('MinIO Node.js - Operações Básicas');
    console.log('='.repeat(60));
    
    const bucketName = 'nodejs-example-bucket';
    
    // 1. Criar bucket
    console.log('\n1. Criando bucket...');
    await createBucket(bucketName);
    
    // 2. Upload de string
    console.log('\n2. Fazendo upload de conteúdo...');
    await uploadString(
        bucketName,
        'hello.txt',
        'Olá MinIO! Este é um teste com Node.js.'
    );
    
    // 3. Upload de arquivo
    console.log('\n3. Criando e fazendo upload de arquivo...');
    const tempFile = '/tmp/nodejs-example.txt';
    fs.writeFileSync(tempFile, 
        'Este é um arquivo de exemplo criado em Node.js.\n' +
        'MinIO é compatível com Amazon S3.\n'
    );
    await uploadFile(bucketName, 'example.txt', tempFile);
    
    // 4. Listar objetos
    console.log('\n4. Listando objetos...');
    await listObjects(bucketName);
    
    // 5. Obter informações do objeto
    console.log('\n5. Obtendo informações do objeto...');
    await getObjectInfo(bucketName, 'example.txt');
    
    // 6. Download de arquivo
    console.log('\n6. Fazendo download...');
    const downloadPath = '/tmp/downloaded-nodejs-example.txt';
    await downloadFile(bucketName, 'example.txt', downloadPath);
    
    const content = fs.readFileSync(downloadPath, 'utf-8');
    console.log(`   Conteúdo baixado:\n   ${content}`);
    
    // 7. Deletar objeto (opcional)
    // console.log('\n7. Removendo objeto...');
    // await deleteObject(bucketName, 'hello.txt');
    
    console.log('\n' + '='.repeat(60));
    console.log('✓ Todos os exemplos foram executados!');
    console.log('='.repeat(60));
}

// Executar
main().catch(err => {
    console.error('Erro na execução:', err);
    process.exit(1);
});
